File: src/main/utils/game-detector.ts
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { homedir } from "node:os";
import { join, normalize } from "node:path";

function checkSteamLibraries(steamPath: string): string {
  const candidates = [join(steamPath, "steamapps", "common", "Mewgenics")];
  const libVdf = join(steamPath, "steamapps", "libraryfolders.vdf");

  if (existsSync(libVdf)) {
    try {
      const content = readFileSync(libVdf, "utf-8");
      for (const line of content.split("\n")) {
        if (!line.toLowerCase().includes('"path"')) {
          continue;
        }
        const parts = line.split('"');
        for (let i = parts.length - 1; i >= 0; i--) {
          const part = parts[i].trim();
          if (
            part &&
            (part.startsWith("/") ||
              part.startsWith("\\") ||
              (part.includes(":") && part.length > 2))
          ) {
            candidates.push(
              join(normalize(part), "steamapps", "common", "Mewgenics")
            );
            break;
          }
        }
      }
    } catch {
      // ignore
    }
  }

  for (const c of candidates) {
    if (existsSync(c)) {
      return c;
    }
  }
  return "";
}

export function autoDetectGameInstall(): string {
  const steamPaths: string[] = [];

  if (process.platform === "win32") {
    steamPaths.push(
      "C:\\Program Files (x86)\\Steam",
      "C:\\Program Files\\Steam",
      join(homedir(), "Steam")
    );
  } else if (process.platform === "darwin") {
    steamPaths.push(
      join(homedir(), "Library", "Application Support", "Steam"),
      join(homedir(), ".steam", "steam")
    );
  } else {
    steamPaths.push(
      join(homedir(), ".local", "share", "Steam"),
      join(homedir(), ".steam", "steam"),
      "/home/deck/.local/share/Steam"
    );
  }

  for (const steamPath of steamPaths) {
    if (existsSync(steamPath)) {
      const result = checkSteamLibraries(steamPath);
      if (result) {
        return result;
      }
    }
  }

  return "";
}

export function detectSteamAppId(gameDir: string): string {
  if (!gameDir) {
    return "";
  }

  try {
    const parts = gameDir.split(/[/\\]/);
    const mewIdx = parts.lastIndexOf("Mewgenics");
    if (mewIdx < 0 || parts[mewIdx - 1] !== "common") {
      return "";
    }
    const steamappsDir = join(...parts.slice(0, mewIdx - 1));

    if (!existsSync(steamappsDir)) {
      return "";
    }

    for (const name of readdirSync(steamappsDir)) {
      if (!name.startsWith("appmanifest_") || !name.endsWith(".acf")) {
        continue;
      }
      try {
        const content = readFileSync(join(steamappsDir, name), "utf-8");
        if (!content.toLowerCase().includes("mewgenics")) {
          continue;
        }
        const match = name.match(/appmanifest_(\d+)\.acf/);
        if (match) {
          return match[1];
        }
        const appIdMatch = content.match(/"appid"\s+"(\d+)"/);
        if (appIdMatch) {
          return appIdMatch[1];
        }
      } catch {
        continue;
      }
    }
  } catch {
    // ignore
  }

  return "";
}

File: .mewtator/app/utils/game_detector.py
import os
import sys
import re
from pathlib import Path


def auto_detect_game_install() -> str:
    if sys.platform == "win32":
        try:
            import winreg
            key = winreg.OpenKey(
                winreg.HKEY_LOCAL_MACHINE,
                r"SOFTWARE\WOW6432Node\Valve\Steam"
            )
            steam_path, _ = winreg.QueryValueEx(key, "InstallPath")
            key.Close()
            
            return _check_steam_libraries(steam_path)
        except Exception:
            pass
    
    elif sys.platform == "darwin":
        steam_paths = [
            Path.home() / "Library" / "Application Support" / "Steam",
            Path("~/.steam/steam").expanduser(),
        ]
        
        for steam_path in steam_paths:
            if steam_path.exists():
                result = _check_steam_libraries(str(steam_path))
                if result:
                    return result
    
    else:
        # Linux / Steam Deck
        steam_paths = [
            Path.home() / ".local" / "share" / "Steam",  # Common Linux / Steam Deck default
            Path.home() / ".steam" / "steam",
            Path("~/.steam/steam").expanduser(),
            Path("/home/deck/.local/share/Steam"),  # Steam Deck specific
        ]
        
        for steam_path in steam_paths:
            if steam_path.exists():
                result = _check_steam_libraries(str(steam_path))
                if result:
                    return result
    
    return ""


def _check_steam_libraries(steam_path: str) -> str:
    candidates = [
        os.path.join(steam_path, "steamapps", "common", "Mewgenics"),
    ]
    
    lib_vdf = os.path.join(steam_path, "steamapps", "libraryfolders.vdf")
    if os.path.exists(lib_vdf):
        try:
            with open(lib_vdf, "r", encoding="utf-8") as f:
                for line in f:
                    if '"path"' in line.lower():
                        # Extract the path value from the VDF line
                        # Format: "path"		"/some/path"
                        parts = line.split('"')
                        # Find the last quoted string which should be the path value
                        for i in range(len(parts) - 1, -1, -1):
                            part = parts[i].strip()
                            if part and (part.startswith('/') or part.startswith('\\') or (':' in part and len(part) > 2)):
                                path = os.path.normpath(part)
                                candidates.append(os.path.join(path, "steamapps", "common", "Mewgenics"))
                                break
        except Exception:
            pass
    
    for c in candidates:
        if os.path.isdir(c):
            return c
    
    return ""


def detect_steam_app_id(game_dir: str) -> str:
    """
    Detect the Steam App ID for Mewgenics by searching for appmanifest files.
    
    Returns:
        Steam App ID as a string, or empty string if not found.
    """
    if not game_dir:
        return ""
    
    try:
        game_path = Path(game_dir)
        if game_path.name == "Mewgenics" and game_path.parent.name == "common":
            steamapps_dir = game_path.parent.parent
            
            for manifest_file in steamapps_dir.glob("appmanifest_*.acf"):
                try:
                    with open(manifest_file, "r", encoding="utf-8") as f:
                        content = f.read()
                        
                        if '"Mewgenics"' in content or '"mewgenics"' in content.lower():
                            match = re.search(r'appmanifest_(\d+)\.acf', manifest_file.name)
                            if match:
                                return match.group(1)
                            
                            app_id_match = re.search(r'"appid"\s+"(\d+)"', content)
                            if app_id_match:
                                return app_id_match.group(1)
                except Exception:
                    continue
    except Exception:
        pass
    
    return ""
